File: packages/agent/src/events/types/message-ops.ts
/**
 * @fileoverview Message Operation Events
 *
 * Events for message modifications like deletion.
 */

import type { EventId } from './branded.js';
import type { BaseEvent } from './base.js';

// =============================================================================
// Message Operations Events
// =============================================================================

/**
 * Message deleted event - soft-deletes a message from context reconstruction
 * The original message event is preserved; this event marks it as deleted.
 * Two-pass reconstruction filters out deleted messages.
 */
export interface MessageDeletedEvent extends BaseEvent {
  type: 'message.deleted';
  payload: {
    /** Event ID of the message being deleted */
    targetEventId: EventId;
    /** Original event type (for validation) */
    targetType: 'message.user' | 'message.assistant' | 'tool.result';
    /** Turn number of deleted message */
    targetTurn?: number;
    /** Reason for deletion */
    reason?: 'user_request' | 'content_policy' | 'context_management';
  };
}

File: packages/agent/src/domains/session/event_store/types/payloads/message_ops.rs
//! Message operation payloads.

use serde::{Deserialize, Serialize};

/// Payload for `message.deleted` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageDeletedPayload {
    /// Event ID of the message being deleted.
    pub target_event_id: String,
    /// Type of the target message.
    pub target_type: String,
    /// Turn number of the deleted message.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_turn: Option<i64>,
    /// Reason for deletion.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}
